lectures/jax_intro.md

```diff
@@ -0,0 +1,594 @@
+
+```{code-cell} ipython3
+a = np.array((2, 1))
+a.sort()
+a
+```
+
+```{code-cell} ipython3
+a = jnp.array((2, 1))
+a.sort()
+a
+```
+
+The designers of JAX chose to make arrays immutable because JAX uses a functional programming style.  More on this below.  
+
+Note that, while mutation is discouraged, it is in fact possible with `at`, as in
+
+```{code-cell} ipython3
+a = jnp.linspace(0, 1, 3)
+id(a)
+```
+
+```{code-cell} ipython3
+a
+```
+
+```{code-cell} ipython3
+a.at[0].set(1)
+```
+
+We can check that the array is mutated by verifying its identity is unchanged:
+
+```{code-cell} ipython3
+id(a)
+```
+
+## Random Numbers
+
+Random numbers are also a bit different in JAX, relative to NumPy.  Typically, in JAX, the state of the random number generator needs to be controlled explicitly.
+
+```{code-cell} ipython3
+import jax.random as random
+```
+
+First we produce a key, which seeds the random number generator.
+
+```{code-cell} ipython3
+key = random.PRNGKey(1)
+```
+
+```{code-cell} ipython3
+type(key)
+```
+
+```{code-cell} ipython3
+print(key)
+```
+
+Now we can use the key to generate some random numbers:
+
+```{code-cell} ipython3
+x = random.normal(key, (3, 3))
+x
+```
+
+If we use the same key again, we initialize at the same seed, so the random numbers are the same:
+
+```{code-cell} ipython3
+random.normal(key, (3, 3))
+```
+
+To produce a (quasi-) independent draw, best practice is to "split" the existing key:
+
+```{code-cell} ipython3
+key, subkey = random.split(key)
+```
+
+```{code-cell} ipython3
+random.normal(key, (3, 3))
+```
+
+```{code-cell} ipython3
+random.normal(subkey, (3, 3))
+```
+
+The function below produces `k` (quasi-) independent random `n x n` matrices using this procedure.
+
+```{code-cell} ipython3
+def gen_random_matrices(key, n, k):
+    matrices = []
+    for _ in range(k):
+        key, subkey = random.split(key)
+        matrices.append(random.uniform(subkey, (n, n)))
+    return matrices
+```
+
+```{code-cell} ipython3
+matrices = gen_random_matrices(key, 2, 2)
+for A in matrices:
+    print(A)
+```
+
+One point to remember is that JAX expects tuples to describe array shapes, even for flat arrays.  Hence, to get a one-dimensional array of normal random draws we use `(len, )` for the shape, as in
+
+```{code-cell} ipython3
+random.normal(key, (5, ))
+```
+
+## JIT Compilation
+
+
+The JAX JIT compiler accelerates logic within functions by fusing linear
+algebra operations into a single, highly optimized kernel that the host can
+launch on the GPU / TPU (or CPU if no accelerator is detected).
+
+
+Consider the following pure Python function.
+
+```{code-cell} ipython3
+def f(x, p=1000):
+    return sum((k*x for k in range(p)))
+```
+
+Let's build an array to call the function on.
+
+```{code-cell} ipython3
+n = 50_000_000
+x = jnp.ones(n)
+```
+
+How long does the function take to execute?
+
+```{code-cell} ipython3
+%time f(x).block_until_ready()
```

Comment:
Thanks, @jstac. This notebook looks good to me and covers almost all the basics to get started with JAX. Just a single suggestion that we can add a line describing the usage of `block_until_ready` from the docs: https://jax.readthedocs.io/en/latest/notebooks/quickstart.html

Reply:
Agreed, thanks @Smit-create .  Perhaps a sentence or two could be added about lazy evaluation.  

In short, JAX tried to bundle tasks for the GPU to maximize efficiency and, in addition, allows the Python interpreter to run ahead of GPU operations.   `block_until_ready` prevents this.  @HumphreyYang , would you mind to add an explanation along these lines?

Reply:
Hi @jstac, 

I think in order to explain what `block_until_ready` is, we should talk a little bit more about how `DeviceArray` works. So I was wondering if I could add a short explanation following this paragraph.

> Even scalar-valued maps on arrays are of type `DeviceArray`:
...
> The term `Device` refers to the hardware accelerator (GPU or TPU), although JAX falls back to the CPU if no accelerator is detected.
> 
> (In the terminology of GPUs, the "host" is the machine that launches GPU operations, while the "device" is the GPU itself.)

Reply:
Please go ahead @HumphreyYang .  Also, would you mind to change my sentence "Even scalar-valued maps on arrays are of type DeviceArray:" to "Even scalar-valued maps on arrays return objects of type DeviceArray:"